### PR TITLE
Ingest add revocation comment

### PR DESCRIPTION
**Pull request by anna-parker**
### Summary

1. Add description of why sequence has been revoked if revoked due to grouping changes
2. Only revoke a sequence due to a regrouping if there is some uncertainty on how to regroup, e.g. if due to new sequences a previous grouping must be split up or two groups have to be joined. If a new segment can be added to an existing group (i.e. was missing before) just add it to the group using a revision
3. Small bug fixes: give revoke_and_regroup cron-job proper permissions to actually be able to do its job
4. Small bug fixes: use the latest `jointAccession` for comparing - otherwise if a new segment is added we can get into a revision loop.

### PR Checklist
Use the new [CCHF nextclade dataset](https://github.com/nextstrain/nextclade_data/pull/265) for testing which improves alignment, resulting in many M segments that were previously lost now being aligned and ingested.

- ingest using current dataset
- set db to persistent
- ingest and processing using new dataset
- run revoke_and_regroup cronjob
- check results

### Screenshots
- Message is sent with sequences that should be revoked and regrouped:
image
- Message after sequence has been revoked is clear:
image

🚀 Preview: https://ingest-add-revocation-com.loculus.org

### CODE REVIEW

**Comment by anna-parker** [comment]
It would be nice to merge this before https://github.com/loculus-project/loculus/pull/2651

File: ingest/scripts/call_loculus.py [hunk 1]
```diff
@@ -218,21 +218,36 @@ def regroup_and_revoke(metadata, sequences, map, config: Config, group_id):
...
+    for item in response:
+        new_accessions[item["submissionId"]] = item["accession"]
 
     to_revoke = json.load(open(map, encoding="utf-8"))
 
-    loc_values = {loc for seq in to_revoke.values() for loc in seq.keys()}
-    loculus_accessions = set(loc_values)
-
-    accessions = {"accessions": list(loculus_accessions)}
+    old_loculus_keys: dict[
```
Comment (theosanderson) [inline ingest/scripts/call_loculus.py:227]: (optional) Above you define `new_accessions` where the values are new accessions, here the values are also `new_accessions` but the name is `old_loculus` (I appreciate the name does contains "keys"!) For me this would be clearer as `old_to_new` and the above as `submission_id_to_new` or something.

File: ingest/scripts/call_loculus.py [hunk 2]
```diff
@@ -218,21 +218,36 @@ def regroup_and_revoke(metadata, sequences, map, config: Config, group_id):
...
+            old_loculus_keys[loc_accession] = all
 
-    response = make_request(HTTPMethod.POST, url, config, json_body=accessions)
-    logger.debug(f"revocation response: {response.json()}")
+    url = f"{organism_url(config)}/revoke"
+    responses = []
+    for old_loc_accession, new_loc_accession in old_loculus_keys.items():
+        logger.debug(f"revoking: {old_loc_accession}")
+        comment = (
+            "INSDC re-ingest found metadata changes, these changes lead the segments in this "
```
Comment (theosanderson) [inline ingest/scripts/call_loculus.py:241]:
```suggestion
            "INSDC re-ingest found metadata changes which lead the segments in this "
```

File: ingest/scripts/call_loculus.py [hunk 3]
```diff
@@ -218,21 +218,36 @@ def regroup_and_revoke(metadata, sequences, map, config: Config, group_id):
...
+            "sequence to be grouped differently, the newly grouped sequences can be found "
```
Comment (theosanderson) [inline ingest/scripts/call_loculus.py:242]:
```suggestion
            "sequence to be grouped differently. The newly grouped sequences can be found "
```

File: ingest/scripts/call_loculus.py [hunk 4]
```diff
@@ -218,21 +218,36 @@ def regroup_and_revoke(metadata, sequences, map, config: Config, group_id):
...
-
-    accessions = {"accessions": list(loculus_accessions)}
+    old_loculus_keys: dict[
+        str, list[str]
+    ] = {}  # Map from old loculus accession to corresponding new accession(s)
+    for key, value in to_revoke.items():
+        for loc_accession in value:
+            all = old_loculus_keys.get(loc_accession, [])
+            all.append(new_accessions[key])
+            old_loculus_keys[loc_accession] = all
```
Comment (theosanderson) [inline ingest/scripts/call_loculus.py:234]: In my understanding this should not be required because two lines above you will get this actual list, not a copy of it - so this has been assigned all along

File: ingest/scripts/call_loculus.py [hunk 5]
```diff
@@ -218,21 +218,36 @@ def regroup_and_revoke(metadata, sequences, map, config: Config, group_id):
...
-    loculus_accessions = set(loc_values)
-
-    accessions = {"accessions": list(loculus_accessions)}
+    old_loculus_keys: dict[
+        str, list[str]
+    ] = {}  # Map from old loculus accession to corresponding new accession(s)
+    for key, value in to_revoke.items():
+        for loc_accession in value:
+            all = old_loculus_keys.get(loc_accession, [])
+            all.append(new_accessions[key])
```
Comment (theosanderson) [inline ingest/scripts/call_loculus.py:233]:
```suggestion
            new_accessions_for_this_old_accession = old_loculus_keys.get(loc_accession, [])
            new_accessions_for_this_old_accession.append(new_accessions[key])
```

File: ingest/scripts/call_loculus.py (hunk 4 above)
Comment (anna-parker) [reply ingest/scripts/call_loculus.py]: I think I need the final line because of the case when it was empty, e.g. [] - there it will not be assigned unless I add it back.

Comment (theosanderson) [reply ingest/scripts/call_loculus.py]: yes! sorry!

**Review by theosanderson — APPROVED** [review]
_Looks_ good to me in terms of readability,etc.! Thanks! In terms of being confident it works:

(A) I presume you have tested it out in some way?
(B) At some point pretty soon we should add code tests for all of this - but it seems mean to do that in this PR specifically :)

**Review by corneliusroemer — COMMENTED** [review]
How is regroup defined? We could in principle not revoke If the change is simply addition of a new segment to an existing segment.

Have you tested revoke with something like: run ingest with sorted authors, then ingest with unsorted authors. That should show some revokes.

**Comment by anna-parker** [comment]
I have only tested with the test examples in the ingest `tests/test_data_cchf` folder. https://github.com/loculus-project/loculus/pull/2651 seems like a good test on actual data. 

At the moment revoke_and_regroup will always revoke the old group and create a new group. I realized handling different cases gets very complicated quickly and as this is such an edge case anyways I didn't think it made sense to go into details yet. That is also why I made revoke_and_regroup only run with manual approval. We can look into the cases and handle them differently if we would like to.

**Comment by anna-parker** [comment]
Result of revocation:

```
15:28:52    DEBUG call_loculus.py:regroup-and-revoke - revocation response: [{'accession': 'LOC_000P2KE', 'version': 2, 'submissionId': 'OK573218.1.L/OK573221.1.M/OK573220.1.S'}]
15:28:52     INFO call_loculus.py:regroup-and-revoke - Revoked: 1 sequence entries of old segment groups
Touching output file results/revoked.
[Tue Apr 29 15:28:52 2025]
Finished jobid: 19 (Rule: regroup_and_revoke)
24 of 25 steps (96%) done
Removing temporary output results/metadata_to_submit_prior_to_revoke_sorted.tsv.
Removing temporary output results/sequences_to_submit_prior_to_revoke_sorted.fasta.
Removing temporary output results/to_revoke.json.
15:29:13     INFO call_loculus.py:approve - Approving sequences
15:29:13     INFO call_loculus.py:approve - Approved: 3 sequences
```
The sequence is nicely marked as revoked:

image

File: kubernetes/loculus/templates/loculus-database-standin.yaml [hunk 6]
```diff
@@ -5,7 +5,7 @@ kind: Deployment
 metadata:
   name: loculus-database
   annotations:
-    argocd.argoproj.io/sync-options: Replace=true
+    argocd.argoproj.io/sync-options: Replace=false
```
Comment (anna-parker) [inline kubernetes/loculus/templates/loculus-database-standin.yaml:8]: To revert before merging:
```suggestion
    argocd.argoproj.io/sync-options: Replace=true
```

File: kubernetes/loculus/templates/loculus-database-standin.yaml [hunk 7]
```diff
@@ -42,6 +42,6 @@ spec:
           value: "trust"
         {{ if not .Values.developmentDatabasePersistence }}
         - name: LOCULUS_VERSION
-          value: {{ $dockerTag }}
+          value: "commit-c6268d7"
```
Comment (anna-parker) [inline kubernetes/loculus/templates/loculus-database-standin.yaml:45]: To revert before merging:
```suggestion
          value: {{ $dockerTag }}
```

File: kubernetes/loculus/values.yaml [hunk 8]
```diff
@@ -1587,8 +1587,8 @@ defaultOrganisms:
         configFile:
           <<: *preprocessingConfigFile
           log_level: DEBUG
-          nextclade_dataset_name: nextstrain/cchfv/linked
-          nextclade_dataset_server: https://raw.githubusercontent.com/nextstrain/nextclade_data/cornelius-cchfv/data_output
+          nextclade_dataset_name: community/pathoplexus/cchfv
+          nextclade_dataset_server: https://raw.githubusercontent.com/anna-parker/nextclade_data/cchfv_updates/data_output
```
Comment (anna-parker) [inline kubernetes/loculus/values.yaml:1591]: To revert before merging
```suggestion
          nextclade_dataset_name: nextstrain/cchfv/linked
          nextclade_dataset_server: https://raw.githubusercontent.com/nextstrain/nextclade_data/cornelius-cchfv/data_output
```

File: kubernetes/loculus/values.yaml [hunk 9]
```diff
@@ -1597,8 +1597,8 @@ defaultOrganisms:
         taxon_id: 3052518
         segment_identification:
           method: "align"
-          nextclade_dataset_server: https://raw.githubusercontent.com/nextstrain/nextclade_data/cornelius-cchfv/data_output
-          nextclade_dataset_name: nextstrain/cchfv/linked
+          nextclade_dataset_server: https://raw.githubusercontent.com/anna-parker/nextclade_data/cchfv_updates/data_output
+          nextclade_dataset_name: community/pathoplexus/cchfv
```
Comment (anna-parker) [inline kubernetes/loculus/values.yaml:1601]: To revert before merging
```suggestion
          nextclade_dataset_server: https://raw.githubusercontent.com/nextstrain/nextclade_data/cornelius-cchfv/data_output
          nextclade_dataset_name: nextstrain/cchfv/linked
```

File: kubernetes/loculus/templates/ingest-deployment.yaml [hunk 10]
```diff
@@ -136,6 +136,21 @@ spec:
...
+                - name: DB_USERNAME
+                  valueFrom:
+                    secretKeyRef:
+                      name: database
+                      key: username
+                - name: DB_PASSWORD
+                  valueFrom:
+                    secretKeyRef:
+                      name: database
+                      key: password
```
Comment (theosanderson) [inline kubernetes/loculus/templates/ingest-deployment.yaml:153]: I don't immediately see a corresponding change in other code?

Comment (theosanderson) [reply kubernetes/loculus/templates/ingest-deployment.yaml:153]: Ah, I'm guessing this just needs a rebase or something

Comment (anna-parker) [reply kubernetes/loculus/templates/ingest-deployment.yaml:153]: this is a small bug that I only now found - the cronjob also needs these keys - they are given to ingest already

File: ingest/scripts/call_loculus.py [hunk 11]
```diff
@@ -352,21 +352,36 @@ def regroup_and_revoke(metadata, sequences, map, config: Config, group_id):
...
-    response = make_request(HTTPMethod.POST, url, config, json_body=accessions)
-    logger.debug(f"revocation response: {response.json()}")
+    url = f"{organism_url(config)}/revoke"
+    responses = []
+    for old_loc_accession, new_loc_accession in old_to_new_loculus_keys.items():
+        logger.debug(f"revoking: {old_loc_accession}")
+        comment = (
+            "INSDC re-ingest found metadata changes which lead the segments in this "
+            "sequence to be grouped differently. The newly grouped sequences can be found "
+            f"here: {" ,".join(new_loc_accession)}."
```
[the remark this answers is not in the document]
Comment (theosanderson) [reply ingest/scripts/call_loculus.py:377]: good catch :)

File: ingest/README.md [hunk 12]
```diff
@@ -103,6 +104,14 @@ We use the Snakemake workflow management system which also uses different config
 
 TLDR: The `Snakefile` contains workflows defined as rules with required input and expected output files. By default Snakemake takes the first rule as the target one and then constructs a graph of dependencies (a DAG) required to produce the expected output of the first rule. The target rule can be specified using `snakemake {rule}`
 
+## Approve Revocations
+
+You might be notified that the ingest pipeline would like to regroup segments of multi-segmented organisms, making the previous grouping obsolete. In this case the old segment-grouping needs to be revoked and the new one added. We do not automate this process yet in case of reingest issues leading to mass revocation of sequences. However, if you approve with the proposed revocation you can run the `regroup_and_revoke` cronjob using:
+
+```
+kubectl create job --from=cronjob/loculus-revoke-and-regroup-cronjob-{config.organism} -n $NAMESPACE loculus-revoke-and-regroup-cronjob-{config.organism}
```
Comment (anna-parker) [inline ingest/README.md:112]: sorry this is invalid it should be:
```suggestion
kubectl create job --from=cronjob/loculus-revoke-and-regroup-cronjob-{config.organism} loculus-revoke-and-regroup-cronjob-{config.organism} -n <NAMESPACE>
```

File: ingest/README.md [hunk 13]
```diff
@@ -56,16 +56,17 @@ We group segments by adding a `jointAccession` field to the metadata which consi
...
 - `submit`: Sequences that have not been submitted before
 - `revise`: Sequences that have been submitted before and have changed
 - `no_change`: Sequences that have been submitted before and have not changed
 - `blocked`: Sequences that have been submitted before but are not in a state that allows updating
+- `revoke`: Only for multi-segmented viruses, these are sequences that were submitted before but have now changed their segment grouping. This means the previously submitted segment-grouping needs to be revoked and the new grouping submitted.
 
 ### Uploading sequences to Loculus
 
-Depending on the triage category, sequences are either submitted as new entries or revised.
+Depending on the triage category, sequences are either submitted as new entries or revised. Furthermore, for multi-segmented organisms where reingest has found grouping changes, maintainers can [trigger](#approve-revocations) the `regroup_and_revoke` rule which revokes the sequences with incorrectly grouped segments and submits sequences with the new segment grouping. We currently do not automate sequence revocation.
```
Comment (corneliusroemer) [inline ingest/README.md:69]:
```suggestion
Depending on the triage category, sequences are either submitted as new entries or revised. Furthermore, for multi-segmented organisms where reingest has found grouping changes, maintainers can [trigger](#approve-revocations) the `regroup_and_revoke` rule which revokes the sequences with incorrectly grouped segments and submits sequences with the new segment grouping. We currently do not fully automate sequence revocation - a human in the loop needs to trigger the job.
```

File: ingest/README.md [hunk 14]
```diff
@@ -103,6 +104,14 @@ We use the Snakemake workflow management system which also uses different config
 
 TLDR: The `Snakefile` contains workflows defined as rules with required input and expected output files. By default Snakemake takes the first rule as the target one and then constructs a graph of dependencies (a DAG) required to produce the expected output of the first rule. The target rule can be specified using `snakemake {rule}`
 
+## Approve Revocations
+
+You might be notified that the ingest pipeline would like to regroup segments of multi-segmented organisms, making the previous grouping obsolete. In this case the old segment-grouping needs to be revoked and the new one added. We do not automate this process yet in case of reingest issues leading to mass revocation of sequences. However, if you approve with the proposed revocation you can run the `regroup_and_revoke` cronjob using:
```
Comment (corneliusroemer) [inline ingest/README.md:109]:
```suggestion
You might be notified that the ingest pipeline would like to regroup segments of multi-segmented organisms, making the previous grouping obsolete. In this case the old segment-grouping needs to be revoked and the new one added. We do not automate this process yet in case of potential reingest bugs leading to erroneous revocation of sequences. However, if you approve with the proposed revocation you can run the `regroup_and_revoke` cronjob using:
```

File: ingest/scripts/compare_hashes.py [hunk 15]
```diff
@@ -127,6 +134,24 @@ def process_hashes(
...
+    if take_subset:
+        return "/".join(
+            f"{record[key]}.{segment}" for key, segment in pairs if record.get(key) in subset
+        )
+
+    return "/".join(f"{record[key]}.{segment}" for key, segment in pairs if record.get(key))
+
+
+def get_last_joint_accession(accession: str, submitted: dict) -> JointInsdcAccession:
+    sorted_versions = sorted(submitted[accession]["versions"], key=lambda x: x["version"])
```
Comment (corneliusroemer) [inline ingest/scripts/compare_hashes.py:150]:
```suggestion
    sorted_versions = sorted(submitted[accession]["versions"], key=lambda x: int(x["version"]))
```

just for safety in case version is not yet an int - otherwise we get issues if we ever reach version 10

File: kubernetes/loculus/templates/ingest-deployment.yaml
Comment (corneliusroemer) [inline kubernetes/loculus/templates/ingest-deployment.yaml:1]: the duplication here between deployment and job that differ only by one argument to the snakemake invocation is a bit suboptimal, but the alternative with helm templating makes things complicated
